### ISSUE TITLE
Type instantiation is excessively deep and possibly infinite

### ISSUE THREAD

**Issue by ClaudiuCeia** [edgedb/edgedb-js#782]
- EdgeDB Version: 4.0+d568f08
- EdgeDB CLI Version: EdgeDB CLI 4.0.2+be1da73
- OS Version: Linux 6.5.6-200.fc38.x86_64 (Fedora 38)
- Deno Version: deno 1.38.0 (release, x86_64-unknown-linux-gnu)
- Typescript Version: typescript 5.2.2

Steps to Reproduce:

1. Try the documented tuple assignment using Deno:

```ts
const spidey = e.tuple({
  name: e.str("Peter Parker"),
  age: e.int64(18)
});
```

2. Observe `Type instantiation is excessively deep and possibly infinite` error:

```
Type instantiation is excessively deep and possibly infinite.deno-ts(2589)
(property) tuple: <{
    name: $expr_Literal<ScalarType<"std::str", string, string, "Peter Parker">>;
    age: $expr_Literal<ScalarType<"std::number", number, number, 18>>;
}>(shape: {
    ...;
}) => $expr_NamedTuple<...> (+3 overloads)
```
 

Schema:

Not applicable, error reproducible for literal named tuple instantiation

---

Cheers, thanks for your hard work, EdgeDB is awesome so far!

**Comment by scotttrinh** [comment]
Very curious that it does not seem to have the same issue with non-Deno TypeScript, even with the same version of TypeScript.

**Comment by ClaudiuCeia** [comment]
A [recurrent theme](https://github.com/edgedb/edgedb-js/issues/462) unfortunately. I guess edgeql-js pushes the boundaries just enough to hit some limitations of the way tsc is integrated into Deno. The weird thing is, I've tried again now with multiple tuple options, including empty tuples and those didn't trigger the error (as expected). But I'm also no longer able to reproduce so I'm wondering if this is maybe related to the way Deno does incremental compilation [which they hint at in the docs](https://docs.deno.com/runtime/manual/advanced/typescript/overview#how-does-it-work).

**Comment by ClaudiuCeia** [comment]
Scrap that - seems to be happening again, not sure what's going on or how I can get it working again.

**Comment by corbinu** [comment]
I see this when I upgrade to Typescript 5.4. Rolling back to 5.3 fixes it

**Comment by scotttrinh** [comment]
Fixed in 1.5.x
